### PR TITLE
Add reduce sparse tensor feature.

### PR DESCRIPTION
**Pull request by gongweibao**

### CODE REVIEW

File: paddle/fluid/framework/details/reduce_op_handle.cc [hunk 1]
```diff
@@ -26,6 +32,112 @@ namespace paddle {
...
+
+  // 1. gather local selected rows, merge them
+  std::string gathered_var_name = out_var_handle->name_ + "_gathered_tmp";
+  auto scope = local_scopes_.at(out_var_handle->scope_idx_);
+  auto gathered_var_mid = scope->Var(gathered_var_name);
+  auto gathered_select_rows =
+      gathered_var_mid->GetMutable<framework::SelectedRows>();
+  GatherLocalSelectedRows(src_selected_rows, in_places, dev_ctxes, out_place,
+                          gathered_select_rows);
+  Wait(dev_ctxes);
```
Comment (typhoonzero) [inline paddle/fluid/framework/details/reduce_op_handle.cc:71]: Why need to wait

Comment (gongweibao) [reply paddle/fluid/framework/details/reduce_op_handle.cc:71]: 拷贝指令是运行在多个卡上的。使用gathered_select_rows前需要等待其他的卡结束。

Comment (typhoonzero) [reply paddle/fluid/framework/details/reduce_op_handle.cc:71]: 这里可以改成全异步，GPU-GPU copy是异步的，GPU->CPU, CPU->GPU使用pin memory也可以异步，可以节省wait的时间

Comment (gongweibao) [reply paddle/fluid/framework/details/reduce_op_handle.cc:71]: 诡异。改成FIXME了。

**Comment by typhoonzero** [comment]
For discussion:

1. can we remove `trainers_endpoints` argument from parallel executor API, or put this config in a struct like `dist_train_specs`
1. Need a unit test for this, or can be add in next PR

**Comment by gongweibao** [comment]
```
can we remove trainers_endpoints argument from parallel executor API, or put this config in a struct like dist_train_specs
```

Hide in program.

File: paddle/fluid/operators/distributed/send_recv.proto.in [hunk 2]
```diff
@@ -28,6 +28,9 @@ service SendRecvService {
   rpc PrefetchVariable(VariableMessage) returns (VariableMessage) {}
 
   rpc CheckpointNotify(VariableMessage) returns (VoidMessage) {}
+
```
Comment (typhoonzero) [inline paddle/fluid/operators/distributed/send_recv.proto.in:31]: please document these two interface

File: python/paddle/fluid/parallel_executor.py [hunk 3]
```diff
@@ -135,9 +135,17 @@ def __init__(self,
...
+        build_strategy.trainer_id = trainer_id
 
         main = main_program
         main = main if main else framework.default_main_program()
+
+        trainers_endpoints = main._trainers_endpoints
+        if num_trainers > 1 and trainers_endpoints:
+            assert num_trainers == len(
+                trainers_endpoints), "num_trainers == len(end_points)"
+            build_strategy.trainers_endpoints = trainers_endpoints
```
Comment (typhoonzero) [inline python/paddle/fluid/parallel_executor.py:147]: This is not used?

**Review by typhoonzero — APPROVED** [review]
things to be update:

1. add test for real sparse collect case
2. fix collect call wait issue
3. use a client singlton instead of context singleton

**Review by qingqing01 — APPROVED** [review]

**Pull request merged**